### ISSUE TITLE
gabc <clear>tag

### ISSUE THREAD

**Issue by jakubjelinek** [gregorio-project/gregorio#1029]
I've noticed that
(c3)
<sp>V/</sp>. R{e}(h)ges(h) Thar(h)sis(h) et(h) ín(hi)su(h)læ(h.) (,)
mú(h')ne(f)ra(e) óf(f_0h)fe(h)rent.(hiH'Ghih.ghG'FE'fggf.) <sp>R/</sp>.(::)
Re(h)ges(h) A(h)ra(h)bum(h) et(h) Sa(hi)ba(h'_) (,)
do(h)na(f') ad(e)dú(f_0h)cent.(hiH'Ghih.ghG'FE'fggf.) (::)

emits R/ too much to the left of the :: bar, haven't checked yet if it is old or new algorithm, but 4.0 wasn't doing this.  Is there an option or some gabc syntax that would tell that I want it only at the bar or later, but not earlier? I've been using R/ etc. on the (::) rather than on the next syllable, which also has ugly issues (if the bar ends up at the end of a line, R/ etc. is emitted there instead of the next line), so perhaps being able to use (::)
<sp>R/</sp>. R{e}(h)...
instead would be nice, but having some option/syntax to say that I don't want it in the middle of previous melisma would be appreciated.  Though it is true Solesmes books put it happily even slightly before the bar.

**Comment by eroux** [comment]
I think you want to tune `maxbaroffsettextX@Y`, see `gsp-default.tex`.

**Comment by jakubjelinek** [comment]
That applies to all text though, right?  I think I'd like different handling for selected special characters, or a way to force it through some gabc markup.  It is fine if normal text is moved to the left, I'd just find it cleaner if verses or subverses couldn't overlap each other.

**Comment by eroux** [comment]
Ok, understood... Well, I don't really know how to do that... perhaps a markup is needed in gabc, I don't really know what it should be... ?

**Comment by henryso** [comment]
How about a new notation in gabc text, let's say `<clear>`, which tells GregorioTeX that the given syllable may not overlap any notes from any previous syllables?  Not that I know how to implement this on the TeX side, but does that sound feasible?

**Comment by rpspringuel** [comment]
One could use the verbatim tags (`<v></v>`) to insert `\hspace` or `\kern` commands before or after the special character, but there would probably be a bit of playing around to get it right.

**Comment by rpspringuel** [comment]
@henryso, I can see two ways of implementing that sort of thing on the TeX side:

1. the `<clear>` signal tells gregorio to emit another syllable type which has different spacing rules built into it.  Probably a bit of overkill, but it would make the clearest distinction.
2. the `<clear>` signal inserts a flag setter (much like we do for the forced centers in the syllable text) which then adjust the behavior of the existing spacing algorithm.  This adds another layer of complexity to that code, but may be the least amount of work.

@jakubjelinek, would this sort of thing be just for bar syllables or would you want to be able to do it with notes as well?

**Comment by henryso** [comment]
If we take approach 1, would we need new syllable types that correspond with `\GreSyllable`, `\GreBarSyllable`, and `\GreNoNoteSyllable`?

**Comment by eroux** [comment]
Yes... I think option `2` is more simple though...

**Comment by henryso** [comment]
I think we were divided evenly between `<clear/>` (@eroux, @jakubjelinek) vs `$clear$` (@henryso, @rpspringuel).  We should resolve that before we start on this.

**Comment by eroux** [comment]
What should be the way to decide that? Maybe a poll on the list?

**Comment by henryso** [comment]
Ok, I'll ask the list.

**Comment by henryso** [comment]
In reference to the previous comment, it was decided that we will support both `<clear>` and `<clear/>`, with `</clear>` being forbidden.

**Comment by henryso** [comment]
After spending a few nights trying to figure out how to adjust the spacing algorithm correctly, I have to wave the flag of surrender.  Should I push the C changes to a branch so someone else can look at the TeX side?

**Comment by rpspringuel** [comment]
If you can push the C changes and a test showing what's wrong I'll try and take a look some time this week.  I do have a test on Wednesday, though, so it probably won't be before then.

Oh, and if you could at lest describe what you've tried on the TeX side, that would also be useful.  Actual non-working code snippets would help too.

**Comment by henryso** [comment]
I'm using the example from the opening post, adding `<clear>` before `R/.`.  The closest I've gotten is to guard lines 1178-1196 and lines 1297-1340 of gregoriotex-syllable.tex (in the fix-1029 branch) with `\ifgre@textcleared\else`.  However, I can't figure out how to extend this to `\GreSyllable`.  Alternately, I tried changing the values of spaces (as indicated earlier by @eroux), but couldn't figure out the right settings.

**Comment by rpspringuel** [comment]
I'm sorry I haven't had time to look at this yet.  School assignments have just been piling on each other for the past week or so and I'm fighting off a cold too so my work output is down as a result.  Assuming I stay healthy, my calendar looks like it might clear our a bit next week.

**Comment by rpspringuel** [comment]
I haven't had time to test anything, but my first instinct after looking this over is to try changing `previousenddifference` to 0 when the clear signal is given.  This should effectively say that there is no way to overlap either the text or the notes because they are to be considered as ending at the same place.  I'd probably do this inside `\gre@calculate@enddifference` because that function currently saves the value in `enddifference` (which applies to the previous syllable) in `previousenddifference` before calculating `enddifference` for the current syllable.  This function is called by all syllable commands, so a change here should be sufficient to effect both bars and notes.

The one complication might be the punctum mora adjustment stuff, but I'd see if the above worked before worrying about that.

**Comment by henryso** [comment]
Using the example from the opening post, here's what I get by setting `previousenddifference` to `0pt` in `\gre@calculate@enddifference`, the punctum mora does have some effect.  That said, with or without a punctum mora, the `R/.` seems off-center under the bar.  I'm not really sure what's wrong or what's right.

![image](https://cloud.githubusercontent.com/assets/421318/20175492/f9012ff4-a710-11e6-848b-44d04dca6d8d.png)

![image](https://cloud.githubusercontent.com/assets/421318/20175455/cb571f50-a710-11e6-9c83-93b0290d8f47.png)

**Comment by rpspringuel** [comment]
What did the result look like before making the change in the code?

Oh, and you could add `\hbox to 0pt{\rule{0.4pt}{12pt}\hss}` both before and after the `R/.` in order to get more visible markers of its horizontal extents.

**Comment by henryso** [comment]
Rolling back the change:

![image](https://cloud.githubusercontent.com/assets/421318/20180431/768b00b4-a728-11e6-883a-f805a7d16e2d.png)

And the same two above, with the hbox:

![image](https://cloud.githubusercontent.com/assets/421318/20180549/eaa58f5a-a728-11e6-8bb7-d0543a03065b.png)

![image](https://cloud.githubusercontent.com/assets/421318/20180683/7d7bee3c-a729-11e6-997b-3dc17e4db9ea.png)

**Comment by rpspringuel** [comment]
Can you push the changes and the test up to GitHub?  I'm going to need to look at debug output to figure out what is going on.

**Comment by henryso** [comment]
Done.  Merged into the fix-1029 branch in both repositories.

**Comment by rpspringuel** [comment]
In looking at this I'm noticing that since the syllable we're working on is at the end of the line the new protrusion code should be active.  However, I don't understand said code well enough to know how that should effect the spacing.  Can someone explain it to me?

**Comment by henryso** [comment]
The protrusion code puts a trailing `.`, `,`, `;`, or `:` (or anything after <pr> in the gabc, of which this has none) in a hbox sized to (1 - protrusion factor)*(normal width).  However, the default protrusion factor of those punctuation signs is 0, so the hbox should be sized to (normal width).

**Comment by rpspringuel** [comment]
No time to try this tonight, but my current line of thinking is that `nextbegindifference` is what's pushing the text out of alignment.  The proof of this would be devising a following syllable with a positive `nextbegindifference` (i.e. notes start before text) and seeing the text in the cleared syllable ending up to the right of the bar instead of the left.

**Comment by rpspringuel** [comment]
Still no time to do any actual work, but I had a brainstorm on other possible tactics for getting this to happen.  Thus I'm recording the ideas here so I don't forget them and incase someone else is inspired by one of them.

Changing `previousenddifference` is intended to make the space which overlaps with previous syllable off-limits when calculating the position of the parts of the cleared syllable.  Since the amount and location of the overlap is available for both bar and note syllables using the same interface (the value of `previousenddifference` this should work for both situations, which is why I originally suggested it.

For bar syllables, however, we might also exploit the offset limits to achieve something similar.  If the offset limits are forced to zero for a cleared syllable (and only that syllable) then the centers of the bar and its text will always be aligned.  This will still allow overlap when one is wider than the other in a way that meshes with the previous syllable (usually when the previous syllable's notes run long and the text is wider than the bar) and will not work if the user asks for a note syllable to be cleared.

Finally, when overlap is allowed and the calculations determine that it's a good idea, there's usually a backwards kern somewhere which effects the overlap.  Eliminating this backwards kern in the case of the cleared syllable might work too.  The problem is that the backwards kern isn't always the first step of printing the syllable because the text is always printed before the notes.  When the notes in the cleared syllable start before the text (an admittedly rare situation) it actually occurs in the middle of the process (after the text has been printed) and is embedded into other shifts happening at the same time.  However, the counteraction always needs to be first.  Thus we would need to calculate it seperately and take that action early.

**Comment by rpspringuel** [comment]
Some more thoughts that I want to record:

The amount of overlap which is allowed is `previousenddifference + begindifference` minus whatever minimum space is required between the syllable elements which are the closest.  In order to prevent overlap, therefore, we have to neutralize both of these sources.  However, we cannot simply neutralize `begindifference` by setting it to 0 the way I've proposed for `previousenddifference` because `begindifference` is also needed to get the two parts of the syllable (the notes and the lyrics) to align properly.  I think, therefore, that we may have to implement the last option in my previous post, in which case it will probably be clearer if we neutralize both elements of the overlap at the same time and in the same place.

**Comment by rpspringuel** [comment]
After some testing I've come up with the following solution:  Right before we start printing the syllable, if the text is to be cleared from the previous one and an overlap would normally be indicated, then skip forward by the smaller of absolute values of `previousenddifference` and `begindifference`.

Question: When clearing a syllable from the previous one, should the last part of the previous syllable abut the first part of the current one or should there be a gap (equal to `syllablefinalskip`) between them?  My solution above leads to the former for bar syllables (which, under the new algorithm, are designed to first neutralize the `syllablefinalskip`) and the latter for notes syllables.  Obviously one needs to be adjusted, I just need to know which one.

Another question: Should I try to apply this to the old bar spacing algorithm or do we want to simply let that atrophy as unsupported going forward?

**Comment by rpspringuel** [comment]
Where is the adjustment for the presence of a punctum mora at the end of 
the prior syllable implemented for notes syllable? 
`\gre@puncutummoraadjustment` is used only for bar syllables and yet 
`\gresetshiftaftermora` indicates that is possible for the mora to be 
ignored by notes syllables too.
-- 
✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝✝
Br. Samuel, OSB
St. Anselm’s Abbey
Washington, DC
(R. Padraic Springuel)

PAX ☧ ΧΡΙΣΤΟΣ

**Comment by rpspringuel** [comment]
I've got something that appears to be most of the way there, at least for bars.  I still need to figure out the handling of the punctum mora adjustment for notes syllables in order to get that working correctly.  Any insight on what I should be looking at?

**Comment by rpspringuel** [comment]
See #1258 for my attempt to get the TeX side working.
